File: MergeSort/CPP/MergeSort.cpp
//
// Created by Shahriar Nasim Nafi on 2/7/22.
//

void merge(int *array, int left, int mid, int right) {

    int leftArraySize = mid - left + 1;

    int rightArraySize = right - mid;

    int sortedIndex = left;

    int leftArray[leftArraySize];
    int rightArray[rightArraySize];

    for (int i = 0; i < leftArraySize; i++) {
        leftArray[i] = array[left + i];
    }

    for (int i = 0; i < rightArraySize; i++) {
        rightArray[i] = array[mid + 1 + i];
    }

    int i = 0, j = 0;

    while (i < leftArraySize && j < rightArraySize) {
        if (leftArray[i] < rightArray[j]) {
            array[sortedIndex++] = leftArray[i++];
        } else if (leftArray[i] > rightArray[j]) {
            array[sortedIndex++] = rightArray[j++];
        } else {
            array[sortedIndex++] = leftArray[i++];
            array[sortedIndex++] = rightArray[j++];
        }
    }
    while (i < leftArraySize) {
        array[sortedIndex++] = leftArray[i++];
    }

    while (j < rightArraySize) {
        array[sortedIndex++] = rightArray[j++];

    }
}

void mergeSort(int *array, int left, int right) {

    if (left >= right) {
        return;
    }

    int mid = left + (right - left) / 2;

    mergeSort(array, left, mid);
    mergeSort(array, mid + 1, right);

    merge(array, left, mid, right);
}

File: MergeSort/Swift/MergeSort.swift
//
// Created by Shahriar Nasim Nafi on 2/7/22.
//

func merge<T: Comparable>(left: [T], right: [T]) -> [T] {
    var merged = [T]()

    var i = 0
    var j = 0

    while i < left.count && j < right.count {
        if left[i] < right[j] {
            merged.append(left[i])
            i+=1

        } else if left[i] > right[j] {
            merged.append(right[j])
            j+=1

        } else {
            merged.append(left[i])
            merged.append(right[j])
            i+=1
            j+=1

        }
    }

    while i < left.count {
        merged.append(left[i])
        i+=1
    }

    while j < right.count {
        merged.append(right[j])
        j+=1
    }

    return merged
}

public extension Array where Element: Comparable {

    func mergeSorted() -> [Element] {
        guard self.count > 1 else { return self }

        let midIndex = self.count / 2

        var left = Array(self[0..<midIndex])
        var right = Array(self[midIndex..<self.count])

        left = left.mergeSorted()
        right = right.mergeSorted()

        return merge(left: left, right: right)
    }
}
